src/Router.js

```diff
@@ -163,21 +152,42 @@ export default class Router {
     let ctx = new Context({ canonicalPath, path, title, params, dispatchId: this.__dispatchId })
 
     if (route) {
-      (replace)
-        ? HistoryEnv.replaceState.apply(null, ctx.getHistoryArgs())
-        : HistoryEnv.pushState.apply(null, ctx.getHistoryArgs())
-      this.__fromPath = this.__currentCanonicalPath
+      if (mode === 'replace') {
+        HistoryEnv.replaceState.apply(null, ctx.getHistoryArgs())
+      } else if (mode === 'push') {
+        HistoryEnv.pushState.apply(null, ctx.getHistoryArgs())
+      }
+
       this.__currentCanonicalPath = canonicalPath
 
-      this.__runHandlers(route.handlers, ctx)
+      this.__runHandlers(route.handlers, ctx, () => {
+        const startTime = this.__startTime;
+        const endTime = fns.getNow();
+        const duration = endTime - startTime;
+        const fromPath = this.__fromPath || 'PAGE LOAD';
+        const toPath = canonicalPath;
+
+        if (this.onRouteComplete) {
+          this.onRouteComplete({
+            fromPath,
+            toPath,
+            duration,
+            startTime,
+            endTime,
+          });
+        }
+
+        this.__fromPath = canonicalPath;
+      })
     } else {
       this.catchall()
     }
   }
 
   __onpopstate(e) {
     if (e.state) {
-      this.replace(e.state.path)
+      this.__dispatch(e.state.path, 'pop');
```

Comment:
yeah the way pop state works is something like this:

```
history: [
  { url: '/v2/projects' }
  { url: '/v2/projects/123/campaigns/456' }
]
```

when you pop state it already does:

```
history: [
  { url: '/v2/projects' }
]
```

Thus the browser history API handles putting us in the right window.history state.  We just need to handle the side-effect of making the page update correctly.